PL2/PL2-core/src/main/java/nl/tudelft/pl2016gr2/core/algorithms/subgraph/SplitGraphs.java

```diff
@@ -74,68 +73,63 @@ public SequenceGraph getSubgraph(Collection<Integer> genomes) {
    * @return the new graph which contains all of the newly created nodes.
    */
-  private SequenceGraph createNewGraph(HashSet<GraphNode> nodeSet, HashSet<Integer> genomeSet) {
-    SequenceGraph newGraph = new HashGraph();
-    nodeSet.forEach(originalNode -> {
-      GraphNode newNode = pruneNode(originalNode, genomeSet, nodeSet);
-      newGraph.add(newNode);
-    });
-    genomeSet.forEach(newGraph::addGenome);
-    return newGraph;
+  private SequenceGraph createNewGraph(HashMap<Integer, GraphNode> oldNodes,
+      HashSet<Integer> genomes) {
+    HashMap<Integer, GraphNode> newNodes = new HashMap<>();
+
+    oldNodes.forEach((id, originalNode) -> newNodes.put(id, originalNode.copy()));
+    HashMap<Integer, GraphNode> prunedNodes = pruneNodes(newNodes, oldNodes, genomes);
+    return new HashGraph(prunedNodes, genomes);
   }
 
   /**
-   * Creates a copy of the <code>original</code> node and prunes irrelevant fields.
+   * Prunes irrelevant fields of the original nodes and applies them to the new nodes.
    * <p>
-   * The method returns a newly instantiated {@link GraphNode} object that is identical, except for
-   * the genomes, in-edges and out-edges.
+   * This method returns the same <code>newNodes</code> map, with all nodes updated correctly.
    * </p>
    * <p>
-   * This operation is performed in <code>O(l+m)</code> for <code>l, m</code> genomes and links
-   * respectively.
+   * This operation is performed in <code>O(n*(k+l+m)</code> for <code>n</code> nodes and <code>k,
+   * l, m</code> genomes, in edges and out edges respectively.
    * </p>
    *
-   * @param original  The original <code>node</code> to prune and copy
-   * @param genomeSet The set of genomes that should be retained in the pruning process
-   * @param nodeSet   The set of nodes that should be retained in the pruning process
-   * @return A new <code>GraphNode</code> identical to the <code>original</code> after pruning
+   * @param newNodes The new nodes to apply the pruned data to
+   * @param oldNodes The original <code>node</code>s whose data to prune
+   * @param genomes  The set of genomes that should be retained in the pruning process
+   * @return The updated map of nodes
    */
-  private GraphNode pruneNode(GraphNode original, HashSet<Integer> genomeSet,
-      HashSet<GraphNode> nodeSet) {
-    GraphNode newNode = original.copy();
-
-    newNode.addAllGenomes(pruneGenomes(original, genomeSet));
-    newNode.addAllInEdges(pruneInLinks(original, nodeSet));
-    newNode.addAllOutEdges(pruneOutLinks(original, nodeSet));
+  private HashMap<Integer, GraphNode> pruneNodes(HashMap<Integer, GraphNode> newNodes,
+      HashMap<Integer, GraphNode> oldNodes, HashSet<Integer> genomes) {
+    newNodes.forEach((id, node) -> {
+      node.addAllInEdges(pruneInEdges(node, newNodes, oldNodes, genomes));
+      node.addAllOutEdges(pruneOutEdges(node, newNodes, oldNodes, genomes));
+      node.addAllGenomes(pruneGenomes(oldNodes.get(id), genomes));
```

Comment:
Shouldn't `set...` be used here instead of `addAll` for all 3 of these?
I think this causes a bug for the genomes, since the copy method coppies the genomes and now you add the genomes to the coppied arraylist, resulting in an arraylist with many duplicates and some non-duplicates which shouldn't have been in the genome list at all.
Note however that the main graph is not altered, since the genome lists are coppied into new arraylist when the node is coppied. Maybe we could change this for some of the lists (if we are sure they aren't altered) so their references are passed (or at least for the `copyAll` methods), since copying all of the lists if they won't be altered is bad for perfomance and memory usage.